test/release/examples/primers/specialMethods.chpl

```diff
@@ -0,0 +1,172 @@
+// primer for more details on initializers and deinitializers.
+
+/*
+  The ``this`` Accessor
+  ---------------------
+*/
+
+// The ``this`` method gives the record the ability to be accessed like an
+// array.  Here we use the index to choose a tuple element.
+proc R.this(n: int) ref {
+  if n < 1 || n > size then
+    halt("index out of bounds accessing R");
+  return vals[n];
+}
+
+var r = new R();
+
+// All functions and methods in Chapel can be called using either parenthesis
+// or square brackets. Here we use square brackets to style it as an array
+// access.
+r[1] = 1;
+r[3] = 3;
+
+writeln(r.vals);
+
+/*
+  Default Iterators
+  -----------------
+*/
+
+// An iterator named ``these`` is automatically called when a record or class
+// instance is used in the iterator position of a ``for`` loop.
+iter R.these() ref {
+  for i in 1..size {
+    yield vals[i];
+  }
+}
+
+for val in r {
+  val += 1;
+}
+writeln(r.vals);
+
+// Classes and records can also define parallel iterators including
+// leader/follower iterator pairs and standalone parallel iterators. For
+// more information on parallel iterators, see the :doc:`parIters` primer.
+
+/*
+  IO Methods
+  ----------
+*/
+
+// Define how to write an instance of R to a channel. We'll write the
+// 'vals' tuple between asterisks. See section
+// :ref:`readThis-writeThis-readWriteThis` for more information  on the
+// ``writeThis``, ``readThis``, and ``readWriteThis`` methods.
+proc R.writeThis(ch: channel) {
+  ch.write("*", vals, "*");
+}
+
+{
+  // Open the file in a new block so that deinitializers
+  // will close it at the end of the block
+  var f = open("tempfile.txt", iomode.cw);
+  var ch = f.writer();
+  ch.writeln(r);
+}
+
+// Define how to read an instance of R from a channel. We'll read the
+// 'vals' tuple between asterisks.
+proc R.readThis(ch: channel) {
+  var star = new ioLiteral("*");
+  ch.read(star);
+  ch.read(vals);
+  ch.read(star);
+}
+
+{
+  var f = open("tempfile.txt", iomode.r);
+  var ch = f.reader();
+  var r2 = new R();
+  ch.readln(r2);
+  assert(r == r2);
+}
+
+// If the record should be read and written using the same
+// format, the combined 'readWriteThis' method can replace the
+// 'readThis' and 'writeThis' methods. This method will be
+// used for both reading and writing the 'vals' tuple
+// surrounded by double asterisks. The ``readThis`` and
+// ``writeThis`` methods defined above have higher precedence
+// than ``readWriteThis``, so this function is not used because
+// they are defined.
+proc R.readWriteThis(ch: channel) {
+  const star = new ioLiteral("**");
+  ch <~> star <~> vals <~> star;
+}
+
+{
+  var chW = openwriter("tempfile.txt");
+  chW.writeln(r);
+  chW.flush();
+
+  writeln(r);
+  var r2 = new R();
+  var chR = openreader("tempfile.txt");
+  chR.readln(r2);
+  assert(r == r2);
+  
+}
+
+/*
+  Operator Overloads
+  ------------------
+*/
+
+// Operators can be overloaded for class and record types to support
```

Comment:
Please make this just about records, since replacing assignment for class instances is probably not something we want to support.